### DISCUSSION TITLE
talking to characters

### DISCUSSION THREAD

**Discussion by AndrejkoL** [Annoraaq/grid-engine#115] [Ideas]
Hi Johannes. Was wondering if you could help me with one more issue. I'm trying to setup interactions between player-player or player-object on key press. So far i tried to get direction which player is facing and then checking if there is another sprite one tile away but i would have to go through all characters position and it just complicated and full of issues. Do you maybe have some easier way of checking if there is another player/object in front of you and executing function between two of them on key press? Any help would be much appreciated as i'm stuck on this for a while now.

**Comment by Annoraaq** [comment]
Hi Andrej,

So far the only way to achieve that would be as you suggested:
1. get neighbouring tile via player position and facing direction
2. call `getAllCharacters()` in order to get the IDs of all characters
3. filter those characters by the position

It could look something like this:

Edit: I corrected the following snippet
```javascript
const facingTilePos = ...;
const charactersOnFacingTile = this.gridEngine.getAllCharacters().filter(charId =>
    this.gridEngine.getPosition(charId).x === facingTilePos.x && this.gridEngine.getPosition(charId).y === facingTilePos.y
);
```

I would not say it is complicated. It is a bit verbose and should be put into it's own function. I could also provide a shorthand in the plugin. But this would only save you from writing the above code. It could be handy but it does not enable you to do more than before.

What do you mean by "full of issues"?

The only problem that I can think of is that in a (probably rare) case where you have a huge amount of characters, it could be slow to iterate over all of them. However, usually the amount of characters is smaller than the amount of tiles in the tilemap. But if that would be a problem, I could also maintain a data structure that stores for each tile position the characters that are currently on that position and give you a constant time lookup.

So if you observe such a case where it slows down your game, let me know and I can optimize it.

**Comment by AndrejkoL** [comment]
I was actually going through array and positions with way more complicated
function that wasn't really working. This is actually so much simpler i
cant believe i havent done it :) Thank you

<https://www.avast.com/sig-email?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail>
Virus-free.
www.avast.com
<https://www.avast.com/sig-email?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail>
<#DAB4FAD8-2DD7-40BB-A1B8-4E2AA1F9FDF2>

**Comment by AndrejkoL** [comment]
Hi Johannes. Tried to do it the way you suggested but still having a few
issues. So my code looks like this:

let facingTilePos;
let playerDirection = this.gridEngine.getFacingDirection("player");
let playerPosition = this.gridEngine.getPosition("player");
facingDir(playerDirection, playerPosition);  //function to determine which
tile im facing

function facingDir(dir, pos){
  if (dir==="up"){
    pos.y-=1;
    facingTilePos = pos;
  }
  else if(dir==="down"){
    pos.y+=1;
    facingTilePos = pos;
  }
  else if(dir==="left"){
    pos.x-=1;
    facingTilePos = pos;
  }
  else if(dir==="right"){
    pos.x+=1;
    facingTilePos = pos;
  }

  return facingTilePos;
}
This part works fine but when i try to filter characters by position it
only gets me an empty array even when there are characters on given tile.
Any suggestions what i'm doing wrong?

const charactersOnFacingTile = this.gridEngine.getAllCharacters().filter(pos =>
    pos.x === facingTilePos.x && pos.y === facingTilePos.y
);

**Reply by Annoraaq** [reply]
Sorry there was a mistake in my above code. You need to get the position for the charId first. I did not test it, though.

```javascript
const charactersOnFacingTile = this.gridEngine.getAllCharacters().filter(charId =>
    this.gridEngine.getPosition(charId).x === facingTilePos.x && this.gridEngine.getPosition(charId).y === facingTilePos.y
);
```

**Comment by AndrejkoL** [comment]
That works perfectly thank you. Ye i realized there was something wrong so
i tried just to getAllCharacters()  first and then use "for function()" to
go through them and compare positions with my facing tile position. Problem
i had was with scope because i couldn't use this.gridEngine.getPosition()
in my separate function because it was undefined. I'm bit new to phaser and
coding in general but is there a way to use this.gridEngine... outside of
create/update? Like in separate function? Thank you

**Reply by Annoraaq** [reply]
Yes. It depends a bit on your code. Phaser initializes the plugin and adds it to the `this`-context that you get in the create/update functions.

Are you calling your function from one of those phaser functions like update or create? In this case you can either pass it as a param:

```javascript
function myFunction(gridEngine) {
  gridEngine.getPosition('player');
  // more gridEngine calls
}
```

Call it like this:
```javascript
function update() {
 // do stuff
 myFunction(this.gridEngine);
}
```

The other way would be to pass the context to the function:

```javascript
function myFunction() {
  this.gridEngine.getPosition('player');
  // more gridEngine calls
}
```

Call it like this:

```javascript
function update() {
 // do stuff
 const myContextBoundFunction = myFunction.bind(this);
 myContextBoundFunction();
}

```

You can also use https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/call or https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/apply

**Comment by AndrejkoL** [comment]
Amazing thanks so much. All works like a charm now.

**Comment by AndrejkoL** [comment]
One more question, is there a way to get character data and properties by
ID? Instead of getAllCharacters(), something like
getCharacterById("player") but with global scope. I want to use custom
properties from characters array for example
 characters: [
      {
        id: "player",
        NPC: true,
        sprite: playerSprite}

**Reply by Annoraaq** [reply]
What do you mean by properties? If you have the ID you can get all relevant properties like the tile position and the facing direction. I could add a method the retrieve the sprite, but since you have to create it outside of the plugin anyway, you already have a reference to it.

If you want to store custom data for your character, you should do it outside the plugin, because it is specific to your game.

For example you could keep a hash map or an object like this:

```javascript
const characters = {
  'player': { NPC: false, sprite: playerSprite, ...},
  'npc1': { NPC: true, sprite: playerSprite, ...},
  ...
 }
```
